File: js/src/encode.ts
/* Internal: encoding primitives (mirrors msgpack_blob_encode.cpp). */

import * as F from "./format.ts";
import { Buf } from "./format.ts";
import { IntWidth, Type, Value } from "./value.ts";

export function encNil(out: Buf): void {
  out.push(F.MP_NIL);
}

export function encBool(out: Buf, v: boolean): void {
  out.push(v ? F.MP_TRUE : F.MP_FALSE);
}

export function encInteger(out: Buf, x: bigint): void {
  if (x >= 0n) {
    if (x <= 0x7fn) out.push(Number(x));
    else if (x <= 0xffn) {
      out.push(F.MP_UINT8);
      out.push(Number(x));
    } else if (x <= 0xffffn) {
      out.push(F.MP_UINT16);
      out.pushU16(Number(x));
    } else if (x <= 0xffffffffn) {
      out.push(F.MP_UINT32);
      out.pushU32(Number(x));
    } else {
      out.push(F.MP_UINT64);
      out.pushU64(x);
    }
  } else {
    if (x >= -32n) out.push(Number(x & 0xffn));
    else if (x >= -128n) {
      out.push(F.MP_INT8);
      out.push(Number(x & 0xffn));
    } else if (x >= -32768n) {
      out.push(F.MP_INT16);
      out.pushU16(Number(x & 0xffffn));
    } else if (x >= -2147483648n) {
      out.push(F.MP_INT32);
      out.pushU32(Number(x & 0xffffffffn));
    } else {
      out.push(F.MP_INT64);
      out.pushU64(x);
    }
  }
}

export function encUnsigned(out: Buf, x: bigint): void {
  x = BigInt.asUintN(64, x);
  if (x <= 0x7fn) out.push(Number(x));
  else if (x <= 0xffn) {
    out.push(F.MP_UINT8);
    out.push(Number(x));
  } else if (x <= 0xffffn) {
    out.push(F.MP_UINT16);
    out.pushU16(Number(x));
  } else if (x <= 0xffffffffn) {
    out.push(F.MP_UINT32);
    out.pushU32(Number(x));
  } else {
    out.push(F.MP_UINT64);
    out.pushU64(x);
  }
}

export function encReal(out: Buf, d: number): void {
  out.push(F.MP_FLOAT64);
  out.pushF64(d);
}

export function encReal32(out: Buf, f: number): void {
  out.push(F.MP_FLOAT32);
  out.pushF32(f);
}

export function encString(out: Buf, s: Uint8Array): void {
  const n = s.length;
  if (n <= 31) out.push(F.MP_FIXSTR_MASK | n);
  else if (n <= 0xff) {
    out.push(F.MP_STR8);
    out.push(n);
  } else if (n <= 0xffff) {
    out.push(F.MP_STR16);
    out.pushU16(n);
  } else {
    out.push(F.MP_STR32);
    out.pushU32(n);
  }
  out.pushBytes(s);
}

export function encBinary(out: Buf, data: Uint8Array): void {
  const n = data.length;
  if (n <= 0xff) {
    out.push(F.MP_BIN8);
    out.push(n);
  } else if (n <= 0xffff) {
    out.push(F.MP_BIN16);
    out.pushU16(n);
  } else {
    out.push(F.MP_BIN32);
    out.pushU32(n);
  }
  out.pushBytes(data);
}

export function encExt(out: Buf, typeCode: number, data: Uint8Array): void {
  const n = data.length;
  if (n === 1) out.push(F.MP_FIXEXT1);
  else if (n === 2) out.push(F.MP_FIXEXT2);
  else if (n === 4) out.push(F.MP_FIXEXT4);
  else if (n === 8) out.push(F.MP_FIXEXT8);
  else if (n === 16) out.push(F.MP_FIXEXT16);
  else if (n <= 0xff) {
    out.push(F.MP_EXT8);
    out.push(n);
  } else if (n <= 0xffff) {
    out.push(F.MP_EXT16);
    out.pushU16(n);
  } else {
    out.push(F.MP_EXT32);
    out.pushU32(n);
  }
  out.push(typeCode & 0xff);
  out.pushBytes(data);
}

export function encInt8(out: Buf, x: bigint): void {
  out.push(F.MP_INT8);
  out.push(Number(BigInt.asUintN(8, x)));
}
export function encInt16(out: Buf, x: bigint): void {
  out.push(F.MP_INT16);
  out.pushU16(Number(BigInt.asUintN(16, x)));
}
export function encInt32(out: Buf, x: bigint): void {
  out.push(F.MP_INT32);
  out.pushU32(Number(BigInt.asUintN(32, x)));
}
export function encInt64(out: Buf, x: bigint): void {
  out.push(F.MP_INT64);
  out.pushU64(x);
}
export function encUint8(out: Buf, x: bigint): void {
  out.push(F.MP_UINT8);
  out.push(Number(BigInt.asUintN(8, x)));
}
export function encUint16(out: Buf, x: bigint): void {
  out.push(F.MP_UINT16);
  out.pushU16(Number(BigInt.asUintN(16, x)));
}
export function encUint32(out: Buf, x: bigint): void {
  out.push(F.MP_UINT32);
  out.pushU32(Number(BigInt.asUintN(32, x)));
}
export function encUint64(out: Buf, x: bigint): void {
  out.push(F.MP_UINT64);
  out.pushU64(x);
}

export function encArrayHeader(out: Buf, count: number): void {
  if (count <= 15) out.push(F.MP_FIXARRAY_MASK | count);
  else if (count <= 0xffff) {
    out.push(F.MP_ARRAY16);
    out.pushU16(count);
  } else {
    out.push(F.MP_ARRAY32);
    out.pushU32(count);
  }
}

export function encMapHeader(out: Buf, count: number): void {
  if (count <= 15) out.push(F.MP_FIXMAP_MASK | count);
  else if (count <= 0xffff) {
    out.push(F.MP_MAP16);
    out.pushU16(count);
  } else {
    out.push(F.MP_MAP32);
    out.pushU32(count);
  }
}

export function encTimestamp(out: Buf, sec: bigint, nsec: number): void {
  if (nsec === 0 && sec >= 0n && sec <= 0xffffffffn) {
    out.push(F.MP_FIXEXT4);
    out.push(0xff);
    out.pushU32(Number(sec));
  } else if (sec >= 0n && sec <= 0x3ffffffffn) {
    out.push(F.MP_FIXEXT8);
    out.push(0xff);
    out.pushU64((BigInt(nsec) << 34n) | sec);
  } else {
    out.push(F.MP_EXT8);
    out.push(12);
    out.push(0xff);
    out.pushU32(nsec >>> 0);
    out.pushU64(sec);
  }
}

export function encodeValue(out: Buf, v: Value): void {
  const t = v.type();
  if (t === Type.Nil) encNil(out);
  else if (t === Type.True) encBool(out, true);
  else if (t === Type.False) encBool(out, false);
  else if (t === Type.Integer) {
    const w = v.intWidth();
    if (w === IntWidth.Int8) encInt8(out, v.asInt64());
    else if (w === IntWidth.Int16) encInt16(out, v.asInt64());
    else if (w === IntWidth.Int32) encInt32(out, v.asInt64());
    else if (w === IntWidth.Int64) encInt64(out, v.asInt64());
    else if (w === IntWidth.Uint8) encUint8(out, v.asUint64());
    else if (w === IntWidth.Uint16) encUint16(out, v.asUint64());
    else if (w === IntWidth.Uint32) encUint32(out, v.asUint64());
    else if (w === IntWidth.Uint64) encUint64(out, v.asUint64());
    else encInteger(out, v.asInt64());
  } else if (t === Type.Real) encReal(out, v.asDouble());
  else if (t === Type.Float32) encReal32(out, v.asFloat());
  else if (t === Type.String) encString(out, v.asBytes());
  else if (t === Type.Binary) encBinary(out, v.blobData());
  else if (t === Type.Ext) encExt(out, v.extType(), v.blobData());
  else if (t === Type.Timestamp) encTimestamp(out, v.timestampSeconds(), v.timestampNanoseconds());
  else encNil(out);
}

File: rust/src/encode.rs
//! Internal: encoding primitives (mirrors `msgpack_blob_encode.cpp`).

use crate::format as f;
use crate::value::{IntWidth, Type, Value};

pub fn enc_nil(out: &mut Vec<u8>) {
    out.push(f::MP_NIL);
}

pub fn enc_bool(out: &mut Vec<u8>, v: bool) {
    out.push(if v { f::MP_TRUE } else { f::MP_FALSE });
}

pub fn enc_integer(out: &mut Vec<u8>, x: i64) {
    if x >= 0 {
        if x <= 0x7f {
            out.push(x as u8);
        } else if x <= 0xff {
            out.push(f::MP_UINT8);
            out.push(x as u8);
        } else if x <= 0xffff {
            out.push(f::MP_UINT16);
            f::push16(out, x as u16);
        } else if x <= 0xffff_ffff {
            out.push(f::MP_UINT32);
            f::push32(out, x as u32);
        } else {
            out.push(f::MP_UINT64);
            f::push64(out, x as u64);
        }
    } else if x >= -32 {
        out.push(x as u8);
    } else if x >= -128 {
        out.push(f::MP_INT8);
        out.push(x as u8);
    } else if x >= -32768 {
        out.push(f::MP_INT16);
        f::push16(out, x as u16);
    } else if x >= -2147483648 {
        out.push(f::MP_INT32);
        f::push32(out, x as u32);
    } else {
        out.push(f::MP_INT64);
        f::push64(out, x as u64);
    }
}

pub fn enc_unsigned(out: &mut Vec<u8>, x: u64) {
    if x <= 0x7f {
        out.push(x as u8);
    } else if x <= 0xff {
        out.push(f::MP_UINT8);
        out.push(x as u8);
    } else if x <= 0xffff {
        out.push(f::MP_UINT16);
        f::push16(out, x as u16);
    } else if x <= 0xffff_ffff {
        out.push(f::MP_UINT32);
        f::push32(out, x as u32);
    } else {
        out.push(f::MP_UINT64);
        f::push64(out, x);
    }
}

pub fn enc_real(out: &mut Vec<u8>, d: f64) {
    out.push(f::MP_FLOAT64);
    f::push64(out, d.to_bits());
}

pub fn enc_real32(out: &mut Vec<u8>, val: f32) {
    out.push(f::MP_FLOAT32);
    f::push32(out, val.to_bits());
}

pub fn enc_string(out: &mut Vec<u8>, s: &[u8]) {
    let n = s.len();
    if n <= 31 {
        out.push(f::MP_FIXSTR_MASK | n as u8);
    } else if n <= 0xff {
        out.push(f::MP_STR8);
        out.push(n as u8);
    } else if n <= 0xffff {
        out.push(f::MP_STR16);
        f::push16(out, n as u16);
    } else {
        out.push(f::MP_STR32);
        f::push32(out, n as u32);
    }
    out.extend_from_slice(s);
}

pub fn enc_binary(out: &mut Vec<u8>, data: &[u8]) {
    let n = data.len();
    if n <= 0xff {
        out.push(f::MP_BIN8);
        out.push(n as u8);
    } else if n <= 0xffff {
        out.push(f::MP_BIN16);
        f::push16(out, n as u16);
    } else {
        out.push(f::MP_BIN32);
        f::push32(out, n as u32);
    }
    out.extend_from_slice(data);
}

pub fn enc_ext(out: &mut Vec<u8>, type_code: i8, data: &[u8]) {
    let n = data.len();
    match n {
        1 => out.push(f::MP_FIXEXT1),
        2 => out.push(f::MP_FIXEXT2),
        4 => out.push(f::MP_FIXEXT4),
        8 => out.push(f::MP_FIXEXT8),
        16 => out.push(f::MP_FIXEXT16),
        _ => {
            if n <= 0xff {
                out.push(f::MP_EXT8);
                out.push(n as u8);
            } else if n <= 0xffff {
                out.push(f::MP_EXT16);
                f::push16(out, n as u16);
            } else {
                out.push(f::MP_EXT32);
                f::push32(out, n as u32);
            }
        }
    }
    out.push(type_code as u8);
    out.extend_from_slice(data);
}

pub fn enc_int8(out: &mut Vec<u8>, x: i64) {
    out.push(f::MP_INT8);
    out.push(x as u8);
}
pub fn enc_int16(out: &mut Vec<u8>, x: i64) {
    out.push(f::MP_INT16);
    f::push16(out, x as u16);
}
pub fn enc_int32(out: &mut Vec<u8>, x: i64) {
    out.push(f::MP_INT32);
    f::push32(out, x as u32);
}
pub fn enc_int64(out: &mut Vec<u8>, x: i64) {
    out.push(f::MP_INT64);
    f::push64(out, x as u64);
}
pub fn enc_uint8(out: &mut Vec<u8>, x: u64) {
    out.push(f::MP_UINT8);
    out.push(x as u8);
}
pub fn enc_uint16(out: &mut Vec<u8>, x: u64) {
    out.push(f::MP_UINT16);
    f::push16(out, x as u16);
}
pub fn enc_uint32(out: &mut Vec<u8>, x: u64) {
    out.push(f::MP_UINT32);
    f::push32(out, x as u32);
}
pub fn enc_uint64(out: &mut Vec<u8>, x: u64) {
    out.push(f::MP_UINT64);
    f::push64(out, x);
}

pub fn enc_array_header(out: &mut Vec<u8>, count: u32) {
    if count <= 15 {
        out.push(f::MP_FIXARRAY_MASK | count as u8);
    } else if count <= 0xffff {
        out.push(f::MP_ARRAY16);
        f::push16(out, count as u16);
    } else {
        out.push(f::MP_ARRAY32);
        f::push32(out, count);
    }
}

pub fn enc_map_header(out: &mut Vec<u8>, count: u32) {
    if count <= 15 {
        out.push(f::MP_FIXMAP_MASK | count as u8);
    } else if count <= 0xffff {
        out.push(f::MP_MAP16);
        f::push16(out, count as u16);
    } else {
        out.push(f::MP_MAP32);
        f::push32(out, count);
    }
}

pub fn enc_timestamp(out: &mut Vec<u8>, sec: i64, nsec: u32) {
    if nsec == 0 && (0..=0xffff_ffff).contains(&sec) {
        out.push(f::MP_FIXEXT4);
        out.push(0xff);
        f::push32(out, sec as u32);
    } else if (0..=0x3_FFFF_FFFF).contains(&sec) {
        out.push(f::MP_FIXEXT8);
        out.push(0xff);
        f::push64(out, ((nsec as u64) << 34) | sec as u64);
    } else {
        out.push(f::MP_EXT8);
        out.push(12);
        out.push(0xff);
        f::push32(out, nsec);
        f::push64(out, sec as u64);
    }
}

pub fn encode_value(out: &mut Vec<u8>, v: &Value) {
    match v.get_type() {
        Type::Nil => enc_nil(out),
        Type::True => enc_bool(out, true),
        Type::False => enc_bool(out, false),
        Type::Integer => match v.int_width() {
            IntWidth::Int8 => enc_int8(out, v.as_i64()),
            IntWidth::Int16 => enc_int16(out, v.as_i64()),
            IntWidth::Int32 => enc_int32(out, v.as_i64()),
            IntWidth::Int64 => enc_int64(out, v.as_i64()),
            IntWidth::Uint8 => enc_uint8(out, v.as_u64()),
            IntWidth::Uint16 => enc_uint16(out, v.as_u64()),
            IntWidth::Uint32 => enc_uint32(out, v.as_u64()),
            IntWidth::Uint64 => enc_uint64(out, v.as_u64()),
            IntWidth::Auto => enc_integer(out, v.as_i64()),
        },
        Type::Real => enc_real(out, v.as_f64()),
        Type::Float32 => enc_real32(out, v.as_f32()),
        Type::String => enc_string(out, v.as_bytes()),
        Type::Binary => enc_binary(out, v.blob_data()),
        Type::Ext => enc_ext(out, v.ext_type(), v.blob_data()),
        Type::Timestamp => enc_timestamp(out, v.timestamp_seconds(), v.timestamp_nanoseconds()),
        // Array / Map are not directly encodable as a scalar Value (the C++
        // reference falls through to nil here).
        Type::Array | Type::Map => enc_nil(out),
    }
}
